Bugfix 2428 python embedding with MET_PYTHON_EXE defined

hsoh-u opened the pull request:
## Expected Differences ##

Same as the bugfix for 11.0

If MET_PYTHON_EXE is defined, the met_point_data is saved as temporary NetCDF file and MET reads the NetCDF file by using compile time python. The log message shows if the temporary file was created.
Without the environment variable MET_PYTHON_EXE 
```
DEBUG 3: Initializing MET compile time python instance: /usr/local/met-python3/bin/python3.8
DEBUG 3: Running MET compile time python instance (/usr/local/met-python3/bin/python3.8) to run user's python script (../../../../scripts/python/read_met_point_obs).
DEBUG 2: Processing 78376 observations at 8465 locations.
```
With the environment variable MET_PYTHON_EXE 
```
DEBUG 3: Initializing MET compile time python instance: /usr/local/met-python3/bin/python3.8
DEBUG 3: Running user-specified python instance (MET_PYTHON_EXE=/usr/local/python3/bin/python3) to run user's python script (../../../../scripts/python/read_met_point_obs).
DEBUG 4: Writing temporary Python point data file:
DEBUG 4:        /usr/local/python3/bin/python3 /d1/personal/hsoh/git/bugfixes/bugfix_2428_python_from_env_2/MET/share/met/wrappers/write_tmp_point_nc.py /dev/shm/hsoh/tmp/tmp_met_nc_11538_0 ../../../../scripts/python/read_met_point_obs /d1/personal/hsoh/MET/test_output/NB/pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc
  python: Python Script:        '/d1/personal/hsoh/git/bugfixes/bugfix_2428_python_from_env_2/MET/share/met/wrappers/write_tmp_point_nc.py'
  python: User Command: '../../../../scripts/python/read_met_point_obs /d1/personal/hsoh/MET/test_output/NB/pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc'
  python: Temporary File:       '/dev/shm/hsoh/tmp/tmp_met_nc_11538_0'
  python: Done python script ../../../../scripts/python/read_met_point_obs took 0:00:00.038250
DEBUG 4: Reading temporary Python point data file: /dev/shm/hsoh/tmp/tmp_met_nc_11538_0
DEBUG 2: Processing 78376 observations at 8465 locations.
```

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

at seneca:

```
export MET_BASE=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_main_v11.0/share/met
export MET_PYTHON_EXE=/usr/local/met-python3/bin/python3.8

/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_main_v11.0/bin/plot_point_obs \
   'PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_main_v11.0/scripts/python/read_met_point_obs.py /d1/personal/hsoh/MET/test_output/NB/pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc' \
  pointdata_numpy.ps -dotsize 2.0 -v 4 | grep -v "Plotting location"

unset MET_PYTHON_EXE

/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_main_v11.0/bin/plot_point_obs \
   'PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_main_v11.0/scripts/python/read_met_point_obs.py /d1/personal/hsoh/MET/test_output/NB/pb2nc/ndas.20120409.t12z.prepbufr.tm00.nc' \
  pointdata_numpy.ps -dotsize 2.0 -v 4 | grep -v "Plotting location"

```
Test from [#2428](https://github.com/dtcenter/MET/issues/2428) 

```
export MET_BASE=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/share/met
export MET_PYTHON_EXE=/usr/local/met-python3/bin/python3.8

/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/bin/plot_point_obs \
  "PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt" \
  test.ps -v 4 | grep -v "Plotting location"
```

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**

- [x] Do these changes include sufficient testing updates? **[Yes]**

Unittest "python_point2grid_pb2nc_TMP_user_python" was added. It should produce the same output with python_point2grid_pb2nc_TMP (pb2nc_TMP.nc  and pb2nc_TMP_user_python.nc).

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

One more output python/pb2nc_TMP_user_python.nc

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

Commit:
#2428 Initial release

Commit:
#2428 Added read_tmp_point_nc.py and write_tmp_point_nc.py

Commit:
#2428 Added met_point_obs_nc.py

Commit:
#2428 Added python_point2grid_pb2nc_TMP_user_python

Commit:
#2428 Changed log message

Commit:
#2428 Moved GP, tmp_nc_base_name, and tmp_nc_file_var_name to python3_util.cc & python3_util.h

Commit:
#2428 Moved nc_tools to met_point_obs_nc.py

Commit:
#2428 Moved nc_point_obs to met_point_obs_nc.py

Commit:
#2428 Added tmp_nc_point_obs

Commit:
#2428 Do not check if MET_PYTHON_EXE is the same as the MET_PYTHON_BIN_EXE

georgemccabe commented on data/wrappers/read_tmp_point_nc.py:
```diff
@@ -0,0 +1,30 @@
+import os
+import sys
+
+met_base_dir = os.getenv('MET_BASE',None)
```
I don't think these 3 lines are necessary since the path is appended using the relative path of the script below.

hsoh-u replied:
MET_BASE was removed at read_tmp_point_nc.py and write_tmp_point_nc.py

georgemccabe commented on data/wrappers/write_tmp_point_nc.py:
```diff
@@ -0,0 +1,59 @@
+import sys
+import importlib.util
+
+met_base_dir = os.getenv('MET_BASE',None)
```
I don't think these 3 lines are necessary since the path is appended using the relative path of the script below.

georgemccabe commented on scripts/python/met_point_obs.py:
```diff
@@ -178,34 +192,108 @@ def get_point_data(self):
 
-    def get_type(self, value):
-        return 'string' if isinstance('str') else type(value)
+        if not self.is_numpy_array(self.hdr_typ):
```
To clean up this logic, I suggest putting the is_numpy_array check inside the convert_to_numpy function. If it is already an numpy array, the function could just return the input. I believe the variables are passed by reference so the function could simply return if it is already a numpy array or modify the value if it is not instead of returning a value. It would be good to double check that this works.

```
    def convert_to_numpy(self, value_list):
        if self.is_numpy_array(value_list):
            return
        value_list = np.array(value_list)
```

Then replace:
```
        if not self.is_numpy_array(self.hdr_typ):
            self.hdr_typ = self.convert_to_numpy(self.hdr_typ)
```
with:
```
        self.convert_to_numpy(self.hdr_typ)
```

hsoh-u replied:
Thank you for the suggestion. I'm not sure it works. Is there a way to pass the reference?
Here are my tests:
```
def set_to_var(local_var):
    local_var = ['111', '222']
    print('local_var', local_var)

aaa = ['bb','cc']
set_to_var(aaa)
print(aaa)
```
result
```
('local_var', ['111', '222'])
['bb', 'cc']
```
The content of test.py:
```
class my_class(object):
   def __init__(self):
       self.my_var = [ '111', 'bbb']

   def update_var(self, var_name):
       print('  input at update_var:', var_name)
       var_name = [' abc', 'def' ]
       print('  updated by update_var:', var_name)

   def do_job(self):
       self.update_var(self.my_var)
       print('  after calling update_var:', self.my_var)


test = my_class()
test.do_job()
```
python3 ./test.py
```
  input at update_var: ['111', 'bbb']
  updated by update_var: [' abc', 'def']
  after calling update_var: ['111', 'bbb']
```

georgemccabe replied:
You are right that it won't work. You are able to change the values of a list or dictionary in a function, but not the entire list itself. Thanks for checking.

```
def set_to_var(local_var):
  local_var[0] = 'a'
  print('local_var', local_var)

aaa = ['bb','cc']
set_to_var(aaa)
print(aaa)
```
Result:
> local_var ['a', 'cc']
> ['a', 'cc']

hsoh-u replied:
Good to know this.

georgemccabe commented on data/wrappers/write_tmp_point_nc.py:
```diff
@@ -0,0 +1,59 @@
+print("{p} User Command:\t".format(p=PROMPT)   + repr(' '.join(sys.argv[2:])))
+print("{p} Temporary File:\t".format(p=PROMPT) + repr(sys.argv[1]))
+
+tmp_filename = sys.argv[1]
```
A lot of this logic looks to be the same as other scripts in the data/wrappers directory. We should consider pulling out duplicate code into a function. This isn't required for this PR and could be done during a bigger refactor of the Python Embedding logic.

georgemccabe reviewed (changes requested):
See inline comments

Commit:
#2428 Do not add MET_BASE/python path

DanielAdriaansen commented:
The test listed in the PR for #2428 does not work for me:
```
dadriaan@seneca:~$ export MET_BASE=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/share/met
dadriaan@seneca:~$ export MET_PYTHON_EXE=/usr/local/met-python3/bin/python3.8
dadriaan@seneca:~$ /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/bin/plot_point_obs   "PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt"   test.ps -v 4 | grep -v "Plotting location"
DEBUG 1: Start plot_point_obs by dadriaan(15368) at 2023-02-23 20:29:00Z  cmd: /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/bin/plot_point_obs PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt test.ps -v 4 
DEBUG 1: Default Config File: /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/share/met/config/PlotPointObsConfig_default
DEBUG 4: parse_conf_sid_list() -> Station ID "sid_inc" list contains 0 entries.
DEBUG 4: parse_conf_sid_list() -> Station ID "sid_exc" list contains 0 entries.
DEBUG 3: Plotting grid definition: Projection: Lat/Lon Nx: 720 Ny: 361 lat_ll: -90.000 lon_ll: -0.000 delta_lat: 0.500 delta_lon: 0.500 wrapLon: true
DEBUG 1: Reading point observation file: PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt
DEBUG 3: Initializing MET compile time python instance: /usr/local/met-python3/bin/python3.8
DEBUG 3: Running user-specified python instance (MET_PYTHON_EXE=/usr/local/met-python3/bin/python3.8) to run user's python script (/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point).
DEBUG 4: Writing temporary Python point data file:
DEBUG 4: 	/usr/local/met-python3/bin/python3.8 /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/share/met/wrappers/write_tmp_point_nc.py /tmp/tmp_met_nc_21694_0 /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt
  python: Python Script:	'/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/share/met/wrappers/write_tmp_point_nc.py'
  python: User Command:	'/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt'
  python: Temporary File:	'/tmp/tmp_met_nc_21694_0'
Python Script:	'/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point'
Input File:	'/d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt'
     point_data: Data Length:	2140
     point_data: Data Type:	<class 'list'>
 met_point_data: Data Type:	<class 'dict'>
  python:  ==PYTHON_ERROR== obs_var_table is empty (string)
DEBUG 4: Reading temporary Python point data file: /tmp/tmp_met_nc_21694_0
ERROR  : 
ERROR  : process_python_point_data -> The obs_var_table is empty. Please check if python input is processed properly
ERROR  : 
```

DanielAdriaansen commented:
> The test listed in the PR for #2428 does not work for me:

The case is `MET_PYTHON_EXE` is set and is equal to `MET_PYTHON_BIN_EXE`, which triggers temporary netcdf write/read.

If I `unset MET_PYTHON_EXE`, then it works, but even with a custom `MET_PYTHON_EXE` that is different than `MET_PYTHON_BIN_EXE` it does not work.

So the error is with the write/read of temporary netCDF file somewhere I think.

Commit:
#2428 Set use_var_id at put_data()

hsoh-u commented:
> > The test listed in the PR for #2438 does not work for me:
> 
> The case is `MET_PYTHON_EXE` is set and is equal to `MET_PYTHON_BIN_EXE`, which triggers temporary netcdf write/read.
> 
> If I `unset MET_PYTHON_EXE`, then it works, but even with a custom `MET_PYTHON_EXE` that is different than `MET_PYTHON_BIN_EXE` it does not work.
> 
> So the error is with the write/read of temporary netCDF file somewhere I think.

Fixed and removed setting from MET_BASE (suggestion from @georgemccabe).
Previous /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2 was renamed and the new build was done at the same place: /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2

DanielAdriaansen commented:
I am still getting the same error:
```
dadriaan@seneca:~$ export MET_PYTHON_EXE=/usr/local/met-python3/bin/python3.8
dadriaan@seneca:~$ /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/bin/plot_point_obs   "PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt"   test.ps -v 4 | grep -v "Plotting location"
DEBUG 1: Start plot_point_obs by dadriaan(15368) at 2023-02-23 22:36:27Z  cmd: /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/bin/plot_point_obs PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt test.ps -v 4 
DEBUG 1: Default Config File: /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/share/met/config/PlotPointObsConfig_default
DEBUG 4: parse_conf_sid_list() -> Station ID "sid_inc" list contains 0 entries.
DEBUG 4: parse_conf_sid_list() -> Station ID "sid_exc" list contains 0 entries.
DEBUG 3: Plotting grid definition: Projection: Lat/Lon Nx: 720 Ny: 361 lat_ll: -90.000 lon_ll: -0.000 delta_lat: 0.500 delta_lon: 0.500 wrapLon: true
DEBUG 1: Reading point observation file: PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt
DEBUG 3: Initializing MET compile time python instance: /usr/local/met-python3/bin/python3.8
DEBUG 3: Running user-specified python instance (MET_PYTHON_EXE=/usr/local/met-python3/bin/python3.8) to run user's python script (/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point).
DEBUG 4: Writing temporary Python point data file:
DEBUG 4: 	/usr/local/met-python3/bin/python3.8 /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/share/met/wrappers/write_tmp_point_nc.py /tmp/tmp_met_nc_15875_0 /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt
  python: Python Script:	'/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/share/met/wrappers/write_tmp_point_nc.py'
  python: User Command:	'/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt'
  python: Temporary File:	'/tmp/tmp_met_nc_15875_0'
Python Script:	'/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point'
Input File:	'/d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt'
     point_data: Data Length:	2140
     point_data: Data Type:	<class 'list'>
 met_point_data: Data Type:	<class 'dict'>
DEBUG 4: Reading temporary Python point data file: /tmp/tmp_met_nc_15875_0
ERROR  : 
ERROR  : process_python_point_data -> The obs_var_table is empty. Please check if python input is processed properly
ERROR  : 
```

Can you double check the command I am running is correct after your fix @hsoh-u ?

Commit:
#2428 Set use_var_id from the global attribute

DanielAdriaansen committed:
Fixed typo.

DanielAdriaansen reviewed (approved):
I performed three tests on seneca:

**TEST 1:** `MET_PYTHON_EXE` unset
```
/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/bin/plot_point_obs   "PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt"   test.ps -v 4 | grep -v "Plotting location"
```

**TEST 2:** `MET_PYTHON_EXE` set to a user Python instance (different from compile-time Python instance)
```
export MET_PYTHON_EXE=/home/dadriaan/.conda/envs/icing/bin/python3
dadriaan@seneca:~$ /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/bin/plot_point_obs   "PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt"   test.ps -v 4 | grep -v "Plotting location"
```

**TEST 3:** `MET_PYTHON_EXE` set to the MET compile-time Python instance
```
export MET_PYTHON_EXE=/usr/local/met-python3/bin/python3.8
dadriaan@seneca:~$ /d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/bin/plot_point_obs   "PYTHON_NUMPY=/d1/personal/hsoh/git/pull_request/MET_bugfix_2428_python_from_env_2/scripts/python/read_ascii_point.py /d1/projects/MET/MET_releases/MET-11.0.0/data/sample_obs/ascii/sample_ascii_obs.txt"   test.ps -v 4 | grep -v "Plotting location"
```
All three behave as I expect:
**TEST 1:** Python embedding uses the Python instance from MET compile-time (**NO** writing/reading temporary file) and correctly reports it in the logging at verbosity = 4
**TEST 2:** Python embedding uses the user installed Python instance (writing/reading temporary file) and correctly reports it in the logging at verbosity = 4
**TEST 3:** Python embedding uses the Python instance from MET compile-time (writing/reading temporary file) and correctly reports it in the logging at verbosity = 4 

In addition, I did not have to set `MET_BASE` to find `met_point_obs.py`, or set `PYTHONPATH`, which is extremely desireable.

I reviewed Appendix F and determined these changes don't require any documentation updates, however I made 1 commit to fix a typo. The documentation for Python embedding will be revisited soon in a separate piece of work.
